### PR TITLE
Bugfix: Don't fork bomb in parallel mode

### PR DESCRIPTION
**Pull request by martin-schulze-vireso**
My system has gone down while trying to measure a parallel run of 4000 tests in a file. After some debugging I noticed a lot of processes piling up.
The issue was that GNU parallel's semaphores were started as background tasks. In this case we won't block until the semaphore run actually gets a free slot and just start the next one.

I restructured the code to take advantage of semaphore's waiting. Unfortunately, that complicates the return code handling.

In the process of this I also updated the output code to print test's results as soon as they are finished (but still in order!).
To debug the parallelism I also needed to add the `--dont-parallelize-across-files` switch, as `parallel --keep-order` waits until a process has finished before the output is printed. This was problematic when the process never finishes due to deadlocks.

As these changes would conflict with #326 I rebased them on top of it. So we either land #326 first and rebase this or land this as is and close #326.

@sublimino Given the risk of crashing the running system with a large enough test file, I would recommend to give this high priority.

- [x] I have reviewed the [Contributor Guidelines][contributor].
- [x] I have reviewed the [Code of Conduct][coc] and agree to abide by it

[contributor]: https://github.com/bats-core/bats-core/blob/master/docs/CONTRIBUTING.md
[coc]:         https://github.com/bats-core/bats-core/blob/master/docs/CODE_OF_CONDUCT.md

### CODE REVIEW

**Comment by martin-schulze-vireso** [comment]
Hm, the busybox installation of tail in the bash containers does not support the `--pid` option. I don't know how representative this is of realworld systems. If we cannot use this, then we might have to fall back to polling if the `$output_dir/exitcode` file exists.

**Comment by martin-schulze-vireso** [comment]
I think the Bash 4.0 test has a spurious timing failure due to load. Please rerun to check that.

File: libexec/bats-core/bats-exec-file [hunk 1]
```diff
@@ -134,64 +134,114 @@ function teardown_file() {
...
+    # get the test's pid
+    test_pid=$(cat "$test_output_dir/pid")
+    # GNU parallel semaphore's subprocess will not be our child, 
+    # we cannot use `wait` but we can `tail --pid`!
+    if ! tail --pid="$test_pid" -f /dev/null 2>/dev/null; then
+      # when tail exits with non null, it does not support --pid 
+      # so we have to fall back to polling
+      until [ -e "$test_output_dir/exitcode" ]
+      do
+        sleep 1
```
Comment (sublimino) [inline libexec/bats-core/bats-exec-file:189]: This looks like it could hang forever if there's a bug somewhere in bats. Should we put a timeout in here? Or logging somehow? Not sure either are good ideas...

File: libexec/bats-core/bats [hunk 2]
```diff
@@ -163,6 +163,9 @@ while [[ "$#" -ne 0 ]]; do
     # suppress stdout (on success)
     parallel --env _ echo ::: 1 >/dev/null || exit 1
     ;;
+  --dont-parallelize-across-files)
```
[the remark this answers is not in the document]
Comment (martin-schulze-vireso) [reply libexec/bats-core/bats:166]: I'd prefer the `--no-parallelize-across-file` which is IMHO clearer, because suite is very ambiguous wording and I can only wonder what a parallel file is.

This was intended as internal debugging aid. If we see this as a valuable option, then we document it, right?

File: libexec/bats-core/bats-exec-file (hunk 1 above)
Comment (martin-schulze-vireso) [reply libexec/bats-core/bats-exec-file:189]: That's a tough one.

Firstly, there are two places where we could block: Waiting for the test to start and waiting for it to stop.

The starting is more under our control, so we could get away with timeouts there. However, the stopping is really hard because we can't decide on a timeout that will work for everyone out there. We might want to make this configurable then.
If we have the pid, we can also try to detect if the test process itself is still running and handle that differently.

**Comment by martin-schulze-vireso** [comment]
>1. [...] is it targeting the 1.2 or 1.3 branches?

I would want to get this in ASAP because it will lead to problems with large test code bases that are run in parallel.

> 2. I have started [...] to experiment with benchmarking

I think this is a good idea to pin down performance regressions.

> 3. thanks [...]

Just cleaning up the mess I left behind ;)

**Comment by sublimino** [comment]
@martin-schulze-vireso I've rebased and signed these commits, and I believe it's ready to merge, will get back to it and merge later/tomorrow

**Comment by martin-schulze-vireso** [comment]
> I believe it's ready to merge, will get back to it and merge later/tomorrow

I'd like to get in at least the timeout on the second wait loop.

**Comment by martin-schulze-vireso** [comment]
I added the timeout for the second wait now.

I really don't like the complexity of this solution. There are so many moving parts and concurrencies that might interact badly and I am constantly working around limitations of parallel (+semaphores).

1. parallel can only enforce the concurrency factor within one call or across a shared semaphore -> use semaphores at the test level
2. semaphores standard output will become interleaved and they don't store the `--results` output correctly in the default filesystem layout -> do it yourself
4. semaphores in foreground mode are blocking -> use background task -> leads to fork bomb or requires solving the same problem as semaphore itself
5. semaphores in background mode don't return the exit code -> store it yourself
6. because of doing everything asynchronously and on our own, we don't know when a test is started/ended or if it is stuck.

So basically, we now do nearly everything ourselves except the actual semaphore exclusion.

If I could make a wish, my design goals for the parallel execution would be:

1. easy to use wrapper that ensures correct parallelization factor
2. get immediate but non-interleaved output

**Comment by sublimino** [comment]
> I really don't like the complexity of this solution.

This is also the fundamental point for me, there's a lot of working around parallel's foibles that make this more difficult to maintain. That said it brings a huge amount of functionality and I'm not averse to the patchset. What do you think is the best way forward @martin-schulze-vireso? Perhaps either:

1. Merge this so it gets tested, and see what happens
2. Preempt the noted issues (especially timeout control) and reimplement this.

I am conscious that 2) is a huge amount of work, and would be happy to test 1). You input welcome : ]

**Comment by martin-schulze-vireso** [comment]
> That said it brings a huge amount of functionality

I'm not sure what you mean with it here? GNU parallel? The parallel mode?

Anyways, It is really unfortunate that we have this situation where the lack of library support pit good features against each other:

Without the setup/teardown_file the parallelization would be considerably easier. With only parallelization across files but not within files, we could do away with the semaphores at the cost of a vastly reduced achievable concurrency factor because there are usually fewer files than tests per file. With only parallelization within files we could also do away with semaphores with probably slightly less reduced concurrency factor, this was my initial implementation without semaphores.

> I am conscious that 2) is a huge amount of work

Well, I am not sure if 1) isn't more work down the line. Then again an implementation of 2) entails a semaphore implementation that would also need some real world exposure to mature.

May hope would be that solution 2) would settle once and not need to be changed after that as opposed to the current implementation which looks so fragile that ongoing development might break it with every other change.

My proposal would be: Getting this in is somewhat time critical. I would set a hard limit like end of week for me to get another solution going. If it does not work out, we can use the current solution. If it does, we go with the new one.

**Comment by sublimino** [comment]
> I'm not sure what you mean with it here? GNU parallel? The parallel mode?

Parallel mode in `bats` : ]

> My proposal would be: Getting this in is somewhat time critical. I would set a hard limit like end of week for me to get another solution going. If it does not work out, we can use the current solution. If it does, we go with the new one.

I'm happy with that. I'm out for work all this weekend so perhaps we can check in again Monday 10th? Or happy to choose a date you suggest :pray:

**Comment by martin-schulze-vireso** [comment]
Okay, I think I have a bash implementation of semaphores now. I probably need some more tests to pin this down but I'd like to get some code review now (e.g. by @sublimino). As an added bonus we now have:

1. a reentrant solution: We can call parallelized bats from parallelized bats without it locking itself up. (However, some tests interfere with each other, so the suite as a whole is not parallelizable.)
2. the semaphore slots are fixed, which leaves open an easy path towards implementing #255 if we should wish to do so.

Some concerns/todos:

- [ ] the waiting/polling is somewhat inefficient and could be replaced by a wait queue+signals, which would even allow for prioritizing which test to run next. Obviously, live feedback would profit from starting the tests in their global order.
- [ ]  the current solution relies on `flock` which was not available on my mac system. For some reason it works on the Travis Mac? Should we rely on this or add an alternative shlock implementation for Mac?
- [ ] it occured several times that the parallel path was broken, while sequential running worked. We should be doing all tests in parallel mode too! However, this requires making the test suite ready for parallelization or turning it off where this is not possible.

**Comment by sublimino** [comment]
Apologies for the latency on this, I won't get to it until this weekend at
least due to other commitments. It's top priority on my bats reviews
though 🙏

**Comment by martin-schulze-vireso** [comment]
@sublimino Do you have an updated ETA?

**Review by sublimino — COMMENTED** [review]
I've had a first deep read, and don't have a great deal to contribute! Is there anything, in particular, you'd like me to stress test? Going to see what I can do with the inner loop/hotpath stuff next time.

File: test/fixtures/suite/test_number/file2.bats [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+}
+
+@test "second test in file 2" {
+    [[ "$BATS_TEST_NUMBER" == 2 ]]
+    [[ "$BATS_SUITE_TEST_NUMBER" == 5 ]]
+}
+
+@test "BATS_TEST_NAMES is per file" {
+    echo "${#BATS_TEST_NAMES[@]}"
+    [[ "${#BATS_TEST_NAMES[@]}" == 3 ]]
```
[the remark this answers is not in the document]
Comment (martin-schulze-vireso) [reply test/fixtures/suite/test_number/file2.bats:17]: You're right. I will fix that and push it to #326. I would propose you review that first and get it landed to reduce the load in this review.

File: test/root.bats [hunk 4]
```diff
@@ -32,6 +33,12 @@ setup() {
   cd - >/dev/null
 }
 
+teardown() {
+  # don't cleanup_tmpdir on each test, this interferes with parallel tests
+  # do the cleanup once for this file
+  test_helper::cleanup_tmpdir "$BATS_TEST_NAME"
```
[the remark this answers is not in the document]
Comment (martin-schulze-vireso) [reply test/root.bats:39]: I don't think that is for me to decide. How would you like to style it?

File: libexec/bats-core/bats-exec-file [hunk 5]
```diff
@@ -134,64 +134,95 @@ function teardown_file() {
...
+
+# prints output from all tests in the order they were started
+# $1 == "blocking": wait for a test to finish before printing
+#    != "blocking": abort printing, when a test has not finished
+bats_forward_output_for_parallel_tests() {
+  local status=0
+  # was the next test already started?
+  while [[ $(( test_number_in_suite_of_last_finished_test + 1 )) -le $test_number_in_suite ]]; do
+    # if we are okay with waiting or if the test has already been finished
+    if [[ "$1" == "blocking" ]] || (! kill -0 "$(cat "$output_folder/$(( test_number_in_suite_of_last_finished_test + 1 ))/pid")" 2>/dev/null); then
```
[the remark this answers is not in the document]
Comment (martin-schulze-vireso) [reply libexec/bats-core/bats-exec-file]: I refactored this part and added comments to explain it better.

**Comment by martin-schulze-vireso** [comment]
> I've had a first deep read, and don't have a great deal to contribute! Is there anything, in particular, you'd like me to stress test? 

I think testing that this behaves correctly in normal usage, e.g. with aborting is the most important thing. I think I noticed a slight performance hit, maybe due to the polling in 1s increments, maybe you can quantify that?

>Going to see what I can do with the inner loop/hotpath stuff next time.

I did not try to optimize for performance here. The kill also was unrelated to this, please have a look at the improved comments for an explanation.

**Comment by sublimino** [comment]
@martin-schulze-vireso I've fixed some (simple) merge conflicts and re-run tests

**Comment by martin-schulze-vireso** [comment]
Should I try to rebase this instead, to see if it fixes the error? I am not sure why AppVeyor is failing.

**Comment by sublimino** [comment]
I retriggered the build in appveyor but it didn't update here, please do rebase @martin-schulze-vireso :thinking:

**Comment by martin-schulze-vireso** [comment]
I rebased it, could you look into merging again please? @sublimino

**Comment by sublimino** [comment]
Love the new functionality in https://github.com/bats-core/bats-core/pull/339/files/626a0db67e0a65ec2accaea59726141d7ad661a4..68fd919cdd3d97e084d61c279faad3e4b2fe9c4f

I appreciate your ongoing work on this @martin-schulze-vireso, I will get it merged in the next two days.

**Comment by sublimino** [comment]
Minor nits (these can wait for another PR):
- `test_helper::cleanup_tmpdir` — we should document or standardise helper function naming
- `man/bats.1` needs an update
- in a perfect world the `sleep 3s` could be reduced, I don't know what race conditions have been seen so will not change it 
 
Previous points:
> the waiting/polling is somewhat inefficient and could be replaced by a wait queue+signals, which would even allow for prioritizing which test to run next. Obviously, live feedback would profit from starting the tests in their global order.

I believe this is tabled for now?

> the current solution relies on flock which was not available on my mac system. For some reason it works on the Travis Mac? Should we rely on this or add an alternative shlock implementation for Mac?

I don't have a Mac to test, a search of https://stackoverflow.com/questions/10526651/mac-os-x-equivalent-of-linux-flock1-command suggests we do need a Mac implementation of this :/

> it occured several times that the parallel path was broken, while sequential running worked. We should be doing all tests in parallel mode too! However, this requires making the test suite ready for parallelization or turning it off where this is not possible.

I would like to see the bats internal test suite passing in parallel too : ] I have wondered if metadata/ENVs in the test file itself could be used to prevent parallel execution. Again we can table for the PR IMHO.

So, pending a discussion/implementation of Mac `flock` questions, this is ready to merge :pray:

**Comment by martin-schulze-vireso** [comment]
> > the waiting/polling is somewhat inefficient and could be replaced by a wait queue+signals, which would even allow for prioritizing which test to run next. Obviously, live feedback would profit from starting the tests in their global order.
> I believe this is tabled for now?

Yes, it is only a performance problem. Now performance is of interest when running in parallel but I don't know how hard the impact will be. I believe the worst case would be many short tests that now get inflated to 1s wait time.

> we do need a Mac implementation of this

 I thought so. I'll see what I can do with shlock which seems to be preinstalled.

> I would like to see the bats internal test suite passing in parallel too

This is already part of the current PR. The travis runs inside the container do both a parallel and a sequential run.

**Comment by sublimino** [comment]
> > we do need a Mac implementation of this
> 
> I thought so. I'll see what I can do with shlock which seems to be preinstalled.

Sorry for the extra legwork. If it makes sense I'm happy to merge this so other PRs can be unblocked and merged today, and live with a parallel-broken `master` for  few days?
 
> > I would like to see the bats internal test suite passing in parallel too
> 
> This is already part of the current PR. The travis runs inside the container do both a parallel and a sequential run.

I can't believe I missed this, and FANTASTIC! Works like a dream for me currently, will hammer it :+1:

**Comment by martin-schulze-vireso** [comment]
Now that I come to think of it, I don't know if I validated that the parallel running is working as expected. Just to sum up here are the remaining todos:

- [ ] shlock on MacOS
- [ ] verify parallel tests run in parallel
- [ ] fix naming convetion

> live with a parallel-broken master

I don't know, it is a small subset of all users but many use the master because the releases stalled for such a long time.

**Comment by martin-schulze-vireso** [comment]
One problem with the parallel tests on travis is that we probably only have one core, so we cannot experience true concurrency. This might hide some race conditions but then again this is not even a bulletproof way of finding those on true multicore execution. The main idea was to detect logic errors in the parallel code paths.

**Comment by martin-schulze-vireso** [comment]
Okay, the issues are dealt with. If you find the naming scheme important, could you rectify that issue? Otherwise this is good to go from my side.

**Comment by sublimino** [comment]
Naming scheme just a nit, let's merge this! Any notes on the next merge to target @martin-schulze-vireso? Otherwise I'll appraise and try to work through (guess?) in a size-based order in case of merge conflicts.

**Review by sublimino — APPROVED** [review]

**Comment by martin-schulze-vireso** [comment]
I'd prioritize: #354 (IMHO similar severity to this), #324 (has waited long enough and should be minor), #337 ("blocks" #345) and #348. Then I'd need to get through the WIP stage on the others and the older PRs probably will have conflicts.

**Comment by sublimino** [comment]
Initial failure for Windows build in https://ci.appveyor.com/project/bats-core/bats-core/builds/35234999

image

Have re-run three times and seen no repeat, so a flake warning but merging anyway.

**Comment by martin-schulze-vireso** [comment]
I think I saw one of those in the testruns on my repo as well. We should monitor of this issue persists.
